iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/resource/memory/InsertNodeMemoryEstimator.java

```diff
@@ -0,0 +1,1111 @@
+    long size = RECOVER_PROGRESS_INDEX_SIZE;
+
+    // Memory calculation in reference type, cannot get exact value, roughly estimate
+    size += REENTRANT_READ_WRITE_LOCK_SIZE;
+    if (progressIndex.getDataNodeId2LocalIndex() != null) {
+      size +=
+          NUM_BYTES_OBJECT_HEADER
+              + progressIndex.getDataNodeId2LocalIndex().size()
+                  * (long) (Integer.BYTES + Long.BYTES);
+    }
+    return size;
+  }
+
+  private static long sizeOfSimpleProgressIndex() {
+    // Memory alignment of basic types and reference types in structures
+    return SIMPLE_PROGRESS_INDEX_SIZE;
+  }
+
+  private static long sizeOfStateProgressIndex(StateProgressIndex progressIndex) {
+    // Memory alignment of basic types and reference types in structures
+    long size = STATE_PROGRESS_INDEX_SIZE;
+
+    // Memory calculation in reference type, cannot get exact value, roughly estimate
+    size += REENTRANT_READ_WRITE_LOCK_SIZE;
+    if (progressIndex.getState() != null) {
+      size += NUM_BYTES_OBJECT_HEADER;
+      for (Map.Entry<String, Binary> entry : progressIndex.getState().entrySet()) {
+        size += sizeOfString(entry.getKey()) + sizeOfBinary(entry.getValue());
+      }
+    }
+    return size;
+  }
+
+  private static long sizeOfTimeWindowStateProgressIndex(
+      TimeWindowStateProgressIndex progressIndex) {
+    // Memory alignment of basic types and reference types in structures
+    long size = TIME_WINDOW_STATE_PROGRESS_INDEX_SIZE;
+
+    // Memory calculation in reference type, cannot get exact value, roughly estimate
+    size += REENTRANT_READ_WRITE_LOCK_SIZE;
+    if (progressIndex.getTimeSeries2TimestampWindowBufferPairMap() != null) {
+      size += NUM_BYTES_OBJECT_HEADER;
+      for (Map.Entry<String, Pair<Long, ByteBuffer>> entry :
+          progressIndex.getTimeSeries2TimestampWindowBufferPairMap().entrySet()) {
+        size += sizeOfString(entry.getKey()) + Long.BYTES * 2;
+      }
+    }
+    return size;
+  }
+
+  // =============================Write==================================
+
+  private static long sizeOfBinary(Binary binary) {
+    if (binary == null) {
+      return 0;
+    }
+    // -----header----
+    // -----ref-------
+    // ---------------
+    // --arrayHeader--
+    // ----values-----
+    return RamUsageEstimator.alignObjectSize(NUM_BYTES_OBJECT_HEADER + NUM_BYTES_OBJECT_REF)
+        + RamUsageEstimator.alignObjectSize(NUM_BYTES_ARRAY_HEADER + binary.getValues().length);
+  }
+
+  private static long sizeOfString(String value) {
+    if (value == null) {
+      return 0;
+    }
+    // -----header----
+    // -----ref-------
+    // ---------------
+    // --arrayHeader--
+    // ----values-----
+    return RamUsageEstimator.alignObjectSize(NUM_BYTES_OBJECT_HEADER + NUM_BYTES_OBJECT_REF)
+        + RamUsageEstimator.alignObjectSize(NUM_BYTES_ARRAY_HEADER + value.length());
+  }
+
+  private static long sizeOfStringArray(final String[] values) {
```

Comment:
Defined in "RamUsageEstimator"

Reply:
This function is only applicable to String class calculations in Java 8 and before Java 8.